src/perception/RcnnMopedModule.cpp

```diff
@@ -0,0 +1,163 @@
+
+  return skeleton;
+}
+
+//=============================================================================
+bool RcnnMopedModule::detectObjects(
+    const aikido::planner::WorldPtr& env,
+    ros::Duration timeout,
+    ros::Time timestamp)
+{
+  bool any_valid = false;
+  
+  tf::TransformListener listener(mNodeHandle);
+  
+  visualization_msgs::MarkerArrayConstPtr marker_message =
+      ros::topic::waitForMessage<visualization_msgs::MarkerArray>(
+          mMarkerTopic, mNodeHandle, timeout);
+
+  // Making sure the Message from MopedLite is non empty
+  if (marker_message == nullptr)
+  {
+    dtwarn << "[RcnnMopedModule::detectObjects] nullptr Marker Message "
+           << mMarkerTopic << std::endl;
+    return false;
+  }
+
+  if (marker_message->markers.size() == 0)
+  {
+    dtwarn << "[RcnnMopedModule::detectObjects] No markers on topic "
+           << mMarkerTopic << std::endl;
+    return false;
+  }
+
+  for (const auto& marker_transform : marker_message->markers)
+  {
+    // Get marker, tag ID, and detection transform name
+    const auto& marker_stamp = marker_transform.header.stamp;
+    const auto& detection_frame = marker_transform.header.frame_id;
+    if (!timestamp.isValid() || marker_stamp < timestamp)
+    {
+      continue;
+    }
+
+    ros::Time t0 = ros::Time(0);
+   
+    tf::StampedTransform transform;
+    try
+    {
+      mTfListener.waitForTransform(
+          mReferenceFrameId, detection_frame, t0, timeout);
+
+      mTfListener.lookupTransform(
+          mReferenceFrameId, detection_frame, t0, transform);
+    }
+    catch (const tf::ExtrapolationException& ex)
+    {
+      dtwarn << "[RcnnMopedModule::detectObjects] TF timestamp is "
+                "out-of-date compared to marker timestamp "
+             << ex.what() << std::endl;
+      continue;
+    }
+
+    // Get orientation of marker
+    Eigen::Isometry3d marker_pose =
+        aikido::perception::convertROSPoseToEigen(marker_transform.pose);
+    // Get the transform as a pose matrix
+    Eigen::Isometry3d frame_pose =
+        aikido::perception::convertStampedTransformToEigen(transform);
+    // Compose to get actual skeleton pose
+    Eigen::Isometry3d skel_pose = frame_pose * marker_pose;
+    Eigen::Isometry3d link_offset = mReferenceLink->getWorldTransform();
+    skel_pose = link_offset * skel_pose;
+
+    std::string skel_name = marker_transform.text;
+    std::string skel_uri;
+    if (!skel_name.compare(0, 5, "water"))
+    {
+      skel_uri = "package://pr_ordata/data/objects/water_bottle.urdf";
+    }
+    else
+    {
+      skel_uri = "package://pr_ordata/data/objects/can.urdf";
+    }
+
+    dart::dynamics::SkeletonPtr objSkeletonPtr =
+        makeBodyFromURDF(skel_uri, skel_pose);
+    objSkeletonPtr->setName(skel_name);
+
+    dart::dynamics::SkeletonPtr env_skeleton = env->getSkeleton(skel_name);
+
+    if (env_skeleton != nullptr)
+    {
+      env->removeSkeleton(env_skeleton);
```

Comment:
This might be related to my question above, but why are we removing this skeleton? It seems that these skeletons would be from a previous call to `detectObjects`, but what if MOPED-lite detected fewer cans than the last call? It seems like the extra cans wouldn't be removed.

Reply:
I also think that it will be better if the perception module decides everything including removing a disappeared object. Now this `RcnnMopedModule` will remove a skeleton with `obj_id` only if it gets "None" in `obj_key`.